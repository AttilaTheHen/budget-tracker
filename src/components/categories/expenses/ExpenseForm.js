import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ExpenseForm extends Component {
  state = {
    editing: false,
    key: null,
    name: '',
    price: ''
  };

  static propTypes = {
    expense: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  handleSubmit = event => {
    event.preventDefault();
    const { key, name, price } = this.state;
    const expense = { name, price };
    if(key) expense.key = key;

    this.props.onComplete(expense);
    this.setState({ key: null, name: '', price: '' });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, name, price } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="name" value={name} onChange={this.handleChange}/>
        <InputControl name="price" value={price} onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = props => (
  <p>
    <label>
      {props.name}:
      <input {...props} required />
    </label>
  </p>
);
 
export default ExpenseForm;