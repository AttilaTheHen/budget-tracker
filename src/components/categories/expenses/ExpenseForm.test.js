import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import ExpenseForm from './ExpenseForm';

describe('Expense Form component', () => {
  it('renders add if no expense prop', () => {
    const handleComplete = jest.fn();
    const wrapper = mount(<ExpenseForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});