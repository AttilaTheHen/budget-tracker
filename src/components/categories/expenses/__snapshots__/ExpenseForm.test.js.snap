// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense Form component renders add if no expense prop 1`] = `
<ExpenseForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="price"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          price
          :
          <input
            name="price"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</ExpenseForm>
`;

exports[`Expense Form component renders edit if there is expense prop 1`] = `
<ExpenseForm
  expense={
    Object {
      "name": "bus",
      "price": 5,
    }
  }
  onCancel={[MockFunction]}
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value="bus"
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value="bus"
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="price"
      onChange={[Function]}
      value={5}
    >
      <p>
        <label>
          price
          :
          <input
            name="price"
            onChange={[Function]}
            required={true}
            value={5}
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</ExpenseForm>
`;
